Add http method support to `module create`

chrisbaldauf opened the pull request:
Before this change, the API Gateway HTTP method was hardcoded to 'GET'. This change
plumbs through support for an HTTP method parameter to be passed on the `module create`
command.

Example usage:
    `$ jaws module create users list --method POST`

The provided HTTP method is validated against the list of methods currently supported by API Gateway.

Meant to address #242 

I also made a pass through the jscs errors in the files that I changed to make them match the spec, so it may be helpful to view individual commits. Please let me know if there is something that I missed.

sixlettervariables commented on lib/commands/module_create.js:
```diff
@@ -46,18 +46,25 @@ module.exports.run = function(JAWS, name, action, runtime, pkgMgr, moduleType) {
   }
 
+  httpMethod = httpMethod.toUpperCase();
+  var supportedHttpMethods = ['DELETE', 'GET', 'HEAD', 'OPTIONS', 'PATCH', 'POST', 'PUT'];
```
Perhaps this should go above with `supportedRuntimes`?

chrisbaldauf replied:
@sixlettervariables I thought about that, but since this is only used here, I thought it would be good to keep it closer to the usage site. I could go either way and would be happy to change it if the team feels strongly.

apaatsio commented on bin/jaws:
```diff
@@ -143,7 +143,9 @@ program
-    .option('-p, --package-manager <pm>', '[create]: Select package manager used when creating awsm for publishing. Valid options: npm')
+    .option('-p, --package-manager <pm>',
+            '[create]: Select package manager used when creating awsm for publishing. Valid options: npm')
+    .option('-m, --method <method>', '[create]: API Gateway HTTP method. E.g. GET, POST, PUT, etc.')
```
Please, add "Default is GET."

apaatsio commented on tests/cli/module_create.js:
```diff
@@ -16,7 +16,7 @@ var config = require('../config'),
     JAWS;
 
-describe('Test "new module" command', function() {
+describe.only('Test "new module" command', function() {
```
Remove .only

chrisbaldauf commented:
Hi all,

I believe that I've addressed all of the PR feedback items. Is there anything else that I can or should do (first time committer to this project). Is there any way to get an ETA on when this might be merged so I can plan my downstream work?

staceymoore commented:
Came upon this pull request when searching for a way to specify method type when creating a module. After looking at the diff, the ability to flag for method is great, but what happens when a single resource has multiple methods? Example: /customers has a POST method and then /customers/{id} has GET, PUT and DELETE methods.

chrisbaldauf commented:
@staceymoore, my understanding is that each method will be its own API Gateway and Lambda function pair. I've structured my project a little differently than the examples in order to account for that fact with a folder per route with subfolders for each method. The primary driver behind this convention was to keep the folder structure flat in order to work with the `jaws dash` command. Based on your example methods, I would run:

```
jaws module create customers post --method POST
jaws module create customers-id get # GET is the default method
jaws module create customers-id put --method PUT
jaws module create customers-id delete --method DELETE
```

Which would create the following directory structure:
```
aws_modules/
  customers/
    get/
      awsm.json # apiGateway.cloudFormation.method: GET
  customers-id/
    delete/
      awsm.json # apiGateway.cloudFormation.method: DELETE
    get/
      awsm.json # apiGateway.cloudFormation.method: GET
    put/
      awsm.json # apiGateway.cloudFormation.method: PUT
```

This starts to get a little more cluttered when you have lengthy routes (e.g. `/customers/{id}/orders/{orderId}`), because JAWS currently puts a 19 character limit on module name. 

This is the best pattern that I've come up with so far that will both play nicely with the `jaws dash` command, keep the source code organized, and make the resulting application "rest-y". I'd be interested to see if others have successfully used other conventions.

BTW, I'm working on a project to template out this directory structure based on an input swagger doc, so if the pattern works and the JAWS projects accepts this and similar changes, that should help to take some of the guess work out of the route structuring.

staceymoore commented:
@chrisbaldauf thanks for the detailed and thoughtful response. :+1:

raptortech-js commented:
This PR is awesome, I hope it gets merged soon.

ac360 commented:
Thanks @chrisbaldauf this is a bit outdated and I apologize this never was merged in.  I'm going to add you to the Framework team regardless, since you contributed some amazing work here.  Thank you for everything :)
